# NumPy Aggregations and Statistics
Task 12 - Day 12

In [1]:
import numpy as np

# rows = students, columns = Maths, Science, English
marks = np.array([
    [78, 82, 65],
    [90, 88, 72],
    [55, 60, 58],
    [67, 70, 80],
    [92, 95, 85],
    [45, 50, 62],
    [73, 68, 70],
    [88, 91, 77],
])

subjects = ["Maths", "Science", "English"]
print(marks.shape)

(8, 3)


## Aggregation functions on a single column (Maths)

In [2]:
maths = marks[:, 0]
print("Maths marks:", maths)

Maths marks: [78 90 55 67 92 45 73 88]


In [3]:
print("Sum:", np.sum(maths))

Sum: 588


In [4]:
print("Mean:", np.mean(maths))

Mean: 73.5


In [5]:
print("Median:", np.median(maths))

Median: 75.5


In [6]:
print("Minimum:", np.min(maths))
print("Maximum:", np.max(maths))

Minimum: 45
Maximum: 92


In [7]:
print("Standard deviation:", np.std(maths))

Standard deviation: 15.976545308670458


In [8]:
print("Variance:", np.var(maths))

Variance: 255.25


In [9]:
# extra aggregations
print("Range (max - min):", np.ptp(maths))
print("75th percentile:", np.percentile(maths, 75))
print("Cumulative sum:", np.cumsum(maths))

Range (max - min): 47
75th percentile: 88.5
Cumulative sum: [ 78 168 223 290 382 427 500 588]


## Using the axis parameter

In [10]:
# axis=0 -> down the rows, one result per column (per subject)
print("Mean per subject:  ", np.mean(marks, axis=0))
print("Median per subject:", np.median(marks, axis=0))
print("Max per subject:   ", np.max(marks, axis=0))
print("Min per subject:   ", np.min(marks, axis=0))
print("Std per subject:   ", np.std(marks, axis=0))

Mean per subject:   [73.5   75.5   71.125]
Median per subject: [75.5 76.  71. ]
Max per subject:    [92 95 85]
Min per subject:    [45 50 58]
Std per subject:    [15.97654531 14.98332406  8.66656651]


In [11]:
# axis=1 -> across the columns, one result per row (per student)
student_avg = np.mean(marks, axis=1)
print("Average per student:", student_avg)

Average per student: [75.         83.33333333 57.66666667 72.33333333 90.66666667 52.33333333
 70.33333333 85.33333333]


## Statistical analysis of the dataset

In [12]:
for i, subject in enumerate(subjects):
    col = marks[:, i]
    print(subject)
    print("  mean  :", round(np.mean(col), 2))
    print("  median:", np.median(col))
    print("  std   :", round(np.std(col), 2))
    print("  range :", np.min(col), "-", np.max(col))

Maths
  mean  : 73.5
  median: 75.5
  std   : 15.98
  range : 45 - 92
Science
  mean  : 75.5
  median: 76.0
  std   : 14.98
  range : 50 - 95
English
  mean  : 71.12
  median: 71.0
  std   : 8.67
  range : 58 - 85


In [13]:
top = np.argmax(student_avg)
bottom = np.argmin(student_avg)
print("Top student index:", top, "| average:", round(student_avg[top], 2))
print("Lowest student index:", bottom, "| average:", round(student_avg[bottom], 2))

Top student index: 4 | average: 90.67
Lowest student index: 5 | average: 52.33


## Why mean and median can differ

In [14]:
skewed = np.array([50, 52, 55, 58, 60, 300])
print("Mean:  ", np.mean(skewed))
print("Median:", np.median(skewed))

Mean:   95.83333333333333
Median: 56.5


## Interpretation of the results

Science has the highest average (75.5), followed by Maths (73.5) and English (71.1). English marks are the most consistent, with a standard deviation of about 8.7, while Maths is the most spread out at about 16.0 — the Maths range runs from 45 to 92, so results there vary far more from student to student. Maths also has a median (75.5) slightly above its mean (73.5), which means a few low scores are pulling the average down. Per student, student 4 leads with an average of about 90.7 and student 5 is lowest at about 52.3. In the small skewed example, one extreme value pushes the mean to 95.8 while the median stays at 56.5, which is why the median is the safer measure of a typical value when outliers are present.

## Interview questions

**What is the difference between mean and median?**
The mean is the sum of all values divided by the count, so every value, including extreme ones, affects it. The median is the middle value once the data is sorted, so it ignores how extreme the outer values are. With outliers the two can differ a lot, as in the skewed example above (mean 95.8, median 56.5), and the median gives a better sense of a typical value.

**What does standard deviation measure?**
How spread out the values are around the mean. A small standard deviation means values cluster close to the average, a large one means they are scattered widely. Variance is the same idea before taking the square root, so standard deviation is in the same units as the data.

**What does the axis parameter do in NumPy?**
It sets the direction an aggregation runs along in a multi-dimensional array. `axis=0` collapses the rows and gives one result per column, `axis=1` collapses the columns and gives one result per row. Without an axis, NumPy aggregates over the whole array and returns a single number.